docs/Internal-Consistency-of-System-Descriptions.md

```diff
@@ -0,0 +1,215 @@
+* The user should be in control of how configuration is applied and which level of abstraction is used depending on the use case.
+
+
+## Concept
+
+This section describes the concept how we address the issue of consistency of the system description and how we provide the necessary tools for the user to handle this data. It has two main goals:
+
+* Make it transparent to the user what is happening and why
+* Give user control to adapt the tool to their needs and use cases
+
+### Duplication of data
+
+With scopes covering the same data from different views, e.g. a configuration file in the config-files scope, and the data extraced from the file in the manifest, some data is duplicated. This gives the full picture and can be useful to the user. So the basic approach is to not try to prevent duplication of data, but make it transparent and controllable what effect the duplication has.
+
+This makes writing scopes easier, because they can do simple extraction of data without having to care about effects on other scopes, provided it is clearly defined how the data is used.
+
+### Sequence of operations
+
+When using Machinery each invocation results in a sequence of operations which are applied to a give set of scopes. The exact sequence and the set of scopes define the result. For an inspection that would be applying global filters, then inspecting the data belonging to the scopes, and storing it to a description. For an image build that would be reading the description, exporting the data for each scope, applying export filters, and running the build.
+
+For inspection the sequence does not matter, it only matters which scopes are used, but the result is independent of what is done when. Scopes never require exclusion of files, not even when the files are read and their data is extracted into the manifest. The duplication of information is handled later.
+
+For export it does matter, if data which is present in the manifest part for a scope is applied first or the same data stored in a file in a file scope is applied first. If the file is applied first and then the manifest, the file is overwritten.
+
+The default sequence is to first apply the files and then apply the data from the manifest. The user can prevent that the manifest data is overwriting the file data by excluding the scope which has the manifest data from the export. This gives full control about the result.
+
+To make it easier for users to achieve commonly required results, it might make sense to exclude some scopes from the export by default.
+
+### Definition of scopes
+
+To make it explicit what dependencies and implications a scope has we add an explicit definition of a scope. This is a YAML file, which defines, what filters to apply, what belongs to the scope and other meta data.
+
+#### Filter definition
+
+An example for a minimal filter definition based on the current functionality is a file `scopes/repositories.yam` with the content:
+
+```yaml
+filters:
+  export:
```

Comment:
Based on our discussion yesterday, do we want to explicitly disallow `import` filters?